### PR TITLE
OBG-220 Encryption

### PR DESCRIPTION
**Pull request by max402**

### CODE REVIEW

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@Validated
+@Configuration
+@ConfigurationProperties(prefix = "facade.encryption")
+public class EncryptionProperties {
+    @NotNull
+    private String providerName;
+    @NotNull
+    private String algorithm;
+
+    private int saltLength;
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java]: Makes sense to add @Min validation

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import de.adorsys.opba.protocol.api.services.EncryptionService;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import org.springframework.stereotype.Service;
+import org.springframework.web.context.annotation.RequestScope;
+
+import java.util.Base64;
+
+@Service
+@RequestScope
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java]: While technically this is correct EncryptionService is service scoped, facade should not know anything about web environment. Suggest to remove Service and RequiredArgsConstructor and provide this as standalone java bean

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/FacadeEncryptionService.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package de.adorsys.opba.protocol.facade.services;
+
+import de.adorsys.opba.protocol.api.services.EncryptionService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class FacadeEncryptionService {
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/FacadeEncryptionService.java]: it probably is FacadeEncryptionServiceFactory

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/SecretKeyServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import java.util.Base64;
+
+import static de.adorsys.opba.protocol.facade.utils.EncryptionUtils.getNewSalt;
+
+@Service
+@RequiredArgsConstructor
+public class SecretKeyServiceImpl implements SecretKeyService {
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/SecretKeyServiceImpl.java]: probably SecretKeyOperations is better name. Service is typically one-purpose java class

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 5]
```diff
@@ -58,38 +79,79 @@
...
+        Optional<ServiceSession> existingSession = serviceSessions.findById(serviceSessionId);
+        if (!existingSession.isPresent()) {
+            throw new RuntimeException("Session not found");
+        }
 
-        if (null != request.getFacadeServiceable().getServiceSessionId()) {
-            session.setId(request.getFacadeServiceable().getServiceSessionId());
+        // existing saved key
+        byte[] secretKey = existingSession.get().getSecretKey();
+        if (!ArrUtils.isEmpty(secretKey)) {
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: secretKey can't be null, it is `@Column(nullable = false)`

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 6]
```diff
@@ -58,38 +79,79 @@
...
+        session.setFintechOkUri(facadeServiceable.getFintechRedirectUrlOk());
+        session.setFintechNokUri(facadeServiceable.getFintechRedirectUrlNok());
+        session.setSecretKey(secretKeyService.encrypt(contextWithKey.get().getKeyDTO().getKey()));
+        session.setAlgo(properties.getAlgorithm());
+        session.setSalt(contextWithKey.get().getKeyDTO().getSalt());
+        session.setIterCount(properties.getIterationCount());
+        return serviceSessions.save(session);
+    }
+
+    private KeyDTO getSessionSecretKey(UUID serviceSessionId, String sessionPassword) {
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: If user passes password in request is should take precedence on what is saved in database

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 7]
```diff
@@ -58,38 +79,79 @@
...
+
+        // new secret key
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: If you need to comment some block of code like 
// new secret key
then it is probably asking for new private method newSecretKey

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/dto/KeyDTO.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package de.adorsys.opba.protocol.facade.dto;
+
+import lombok.Value;
+
+@Value
+public class KeyDTO {
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/dto/KeyDTO.java]: Please use KeyDto
no sense to highlight DTO that much

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 9]
```diff
@@ -58,38 +79,79 @@
...
+        if (null == serviceSessionId) {
+            if (Strings.isNullOrEmpty(sessionPassword)) {
+                throw new RuntimeException("No password. Can't generate secret key");
             }
+            byte[] salt = getNewSalt(properties.getSaltLength());
+            byte[] key = secretKeyService.generateKey(sessionPassword, salt);
+            return new KeyDTO(key, salt);
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: Given the importance of encryption, I suggest creating:
KeyDto that simply stores key bytes[]
that extends KeyDto
KeyAndSaltDto that store both key and salt bytes

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 10]
```diff
@@ -58,38 +79,79 @@
...
+    ) {
+        FacadeServiceableRequest facadeServiceable = request.getFacadeServiceable();
+        UUID serviceSessionId = facadeServiceable.getServiceSessionId();
+
+        ServiceSession session = new ServiceSession();
+        if (null != serviceSessionId) {
+            session.setId(serviceSessionId);
+        }
+
+        session.setContext(contextWithKey.get().getEncryptedContext());
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: ServiceSession requires decrypted context

File: opba-db/src/main/resources/migration/migrations/0000-init-schema.xml [hunk 11]
```diff
@@ -32,6 +32,7 @@
             <column name="protocol_bean_name" type="VARCHAR(32)">
                 <constraints nullable="false"/>
             </column>
+            <column name="parent_id" type="BIGINT"/>
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-init-schema.xml]: Not necessary field

File: opba-db/src/main/resources/migration/migrations/0000-init-schema.xml [hunk 12]
```diff
@@ -76,6 +77,10 @@
             <column name="version" type="INT">
                 <constraints nullable="false"/>
             </column>
+            <column name="secret_key" type="BLOB"/>
```
Comment (valb3r) [inline opba-db/src/main/resources/migration/migrations/0000-init-schema.xml]: These columns are non-null

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java (hunk 1 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java]: done

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    private int iterationCount;
```
[the remark this answers is not in the document]
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/config/EncryptionProperties.java]: done

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 14]
```diff
@@ -58,38 +79,79 @@
...
 
     @NotNull
     @SneakyThrows
-    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(T request) {
-        if (null != request.getFacadeServiceable().getServiceSessionId()) {
-            Optional<ServiceSession> existingSession = serviceSessions
-                    .findById(request.getFacadeServiceable().getServiceSessionId());
-            if (existingSession.isPresent()) {
-                return existingSession.get();
+    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: This method always creates new ServiceSession, but it has name `findServiceSessionByIdOrCreate` is it correct?

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java [hunk 15]
```diff
@@ -31,16 +41,27 @@
...
     public <T extends FacadeServiceableGetter> ServiceContext<T> provide(T request) {
-        if (null == request.getFacadeServiceable()) {
+        FacadeServiceableRequest facadeServiceable = request.getFacadeServiceable();
+        if (null == facadeServiceable) {
             throw new IllegalArgumentException("No serviceable body");
         }
 
+        UUID serviceSessionId = facadeServiceable.getServiceSessionId();
+        String sessionPassword = facadeServiceable.getSessionPassword();
+        KeyDTO keyDTO = getSessionSecretKey(serviceSessionId, sessionPassword);
```
Comment (valb3r) [inline opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: This method should be in `extractOrCreateServiceSession`

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/FacadeEncryptionService.java (hunk 3 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/FacadeEncryptionService.java]: renamed

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/SecretKeyServiceImpl.java (hunk 4 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/SecretKeyServiceImpl.java]: renamed

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 5 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: How then we can clear secretKey in db after some time?

Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: that means '@Column(nullable = false)' should be removed

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java (hunk 2 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java]: why remove RequiredArgsConstructor?

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 7 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: ok

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/dto/KeyDTO.java (hunk 8 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/dto/KeyDTO.java]: ok

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 6 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: ok

File: opba-db/src/main/resources/migration/migrations/0000-init-schema.xml (hunk 12 above)
Comment (max402) [reply opba-db/src/main/resources/migration/migrations/0000-init-schema.xml]: ok

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 14 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: I split method but the name remains

Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: renamed to createServiceSession

File: opba-db/src/main/resources/migration/migrations/0000-init-schema.xml (hunk 11 above)
Comment (max402) [reply opba-db/src/main/resources/migration/migrations/0000-init-schema.xml]: removed

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 9 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: done

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 15 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: ok

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 10 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: in this place I persist encrypted session

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java (hunk 2 above)
Comment (valb3r) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java]: @Service
@RequestScope
sorry, remove these

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java (hunk 5 above)
Comment (valb3r) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java]: Use some clear indication of column state - either column is empty then it is NULL or column is empty - then it is empty byte array

File: opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java (hunk 2 above)
Comment (max402) [reply opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/EncryptionServiceImpl.java]: done

**Pull request merged**